### ISSUE TITLE
about feeding word2vec embedding to item features 

### ISSUE THREAD

**Issue by tianpunchh** [lyst/lightfm#331]
Hello,

I think it is a pretty nice functionality that lightfm can feed in some feature dimension for either item or user to start with. This hybrid method is awesome. However, I have a question that hopefully can get some insight from your advice.

My understanding of the workflow is, in Dataset.build_user_features(item_features), it will build up a feature weight for each feature, say features are ['f1', 'f2', 'f3'], and for a particular item the weight could be binary vectors [1, 0, 1], and the algorithm will learn the n_component embedding of each feature: e_f1, e_f2 and e_f3. The final item embedding takes the dot product between [1,0,1] and [e_f1, e_f2, e_f3]  

This is good, for example features are some tags and basically the feature weight here is one-hot vectors to turn on/off tags according to a particular item's property, and a by-product is the embedding of each tag. However, what if I have already had the embedding for an item. From like word2vec or tf-idf, I can get a dense vector embedding for a book item or movie. Say from word2vec an item has embedding [0.3, 0.5, -0.1, -0.9]. I can interpret it as four features with weights 0.3, 0.5, -0.1 and -0.9 (essentially I believe the embedding is to project to a multi-dim feature space, now the meaning of tags are not human understandable tags), so that is fine. Then lightfm will try to embed for each of the word2vec embedding element, and regard the word2vec embedding as weight. Will this word2vec embedding gain some benefit for this case as this looks like a two-step embedding?

And want to confirm that: if I do in this way with word2vec embedding as the weights, there will be some negative weights, I bet lightfm would happily take negative weights without any problem. 

Thank you so much !
Tian

**Comment by maciejkula** [comment]
Hint: you single backticks to provide inline code formatting.

Your understanding of features and the embedding process is correct. You can provide latent embeddings instead of binary features if you like; I don't really have an expectation if this will help or hinder the model. Try and let me know! And yes, negative feature weights are perfectly fine.

**Comment by tianpunchh** [comment]
Hi Sir,

I tried word embedding, it speeds up the training significantly. By using the same learning rate, to converge to the same in sample and out of sample benchmark, word embedding one takes about only 1/4 or 1/3 training epochs compared with that w/o providing features. However, the final result as long as it converges looks similar. I think lightfm indeed performs well to learn embedding itself. All these make sense to me.

BTW, I have a question, does lightfm have an api (or we can easily get it) for this task which I think is natural? Given a user that is involved in the training (basically the model has his/her latent embedding), if we have a new item that is not seen by the model before, but we know its features just as those in the training data (for example, I have its word embedding that is trained in the same way as other items), how can I predict the affinity score of this item to this user? For an already-seen item, we can use model.predict(uid, item_id) since both are registered. For this particular problem,  we have user_latent_embedding and of course we can derive item_latent_embedding, my question is: is the "score" in the lightfm just the inner product between user and item embeddings? Does lightfm have some convenient way to accomplish this task?

**Comment by SimonCW** [comment]
I’m closing this issue because it has been inactive for a long time. If you still encounter the problem, please open a new issue.

Thank you!
